### DISCUSSION TITLE
an-introduction-to-the-heap-data-structure-and-dotnets-priority-queue/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#412] [Announcements]
# an-introduction-to-the-heap-data-structure-and-dotnets-priority-queue/

In this post I provide an introduction to the heap data structure, describe why it's useful, and show how it's used in .NET's PriorityQueue type.

https://andrewlock.net/an-introduction-to-the-heap-data-structure-and-dotnets-priority-queue/

**Comment by jamescurran** [comment]
It's not exactly true that the data structure Heap is "completely unrelated" to the memory-management Heap.

In the distant past days of C, the programs allocated memory thru the run-time library function `malloc`, which used a Heap data structure to keep track of the allocated blocks. Which is why today will still call allocated memory "on the Heap".  (Similarly C compilers put local variables on the CPU's stack, and we still use that expression today too.)

**Reply by dodexahedron** [reply]
I would bet he's pretty familiar with the concept and that it was just a poor choice of words when oversimplifying, since it's still all just memory and the underlying concept of a heap is the same. I think he was just trying to emphasize were discussing logical structures in the runtime, not their underlying implementation below that.

**Comment by qbit86** [comment]
> // The following comparer turns the priority queue into a
> // max-heap instead of min-heap by reversing the comparer.
> var inverseComparer = Comparer<int>.Create((a, b) => 0 - a.CompareTo(b));

You can implement the inverse comparer more easily:
```cs
var inverseComparer = Comparer<int>.Create((a, b) => b.CompareTo(a));
```

**Reply by andrewlock** [reply]
True!

**Comment by softwareliberationarmy** [comment]
Great post, and very timely for a project I'm working on. 

Is the PriorityQueue thread-safe? Or is there an alternate thread-safe version of this type?

**Reply by andrewlock** [reply]
Glad you like it - no, it's not thread safe, and afaik there is no built-in thread safe implementation I'm afraid

**Comment by danielabbatt** [comment]
Small typo or cousings which should be cousins

**Comment by JoaoCampos07** [comment]
Nice article, Andrew. I still remember the days were people were asking for this data structure to be present in .NET. 
After some time, and with the help of the community, there it is, hopefully it will also be thread safe.  :)
